File: tests/differential/python_ts_parity/build_scene.py
#!/usr/bin/env python3
"""Emit canonical JSON scene description for parity diff (sorted keys)."""

from __future__ import annotations

import argparse
import hashlib
import json
import sys
from pathlib import Path


def scene_from_fixture(path: Path) -> dict:
    data = json.loads(path.read_text(encoding="utf-8"))
    if data.get("version") != 1:
        raise ValueError("unsupported fixture version")
    # Canonical numeric payload for scatter / heatmap (deterministic PRNG placeholder).
    layers_out = []
    for layer in data["layers"]:
        t = layer["type"]
        if t == "scatter":
            seed = int(layer["seed"])
            n = int(layer["points"])
            h = hashlib.sha256(f"scatter-{seed}-{n}".encode()).hexdigest()[:16]
            layers_out.append({"type": "scatter", "digest": h, "n": n, "seed": seed})
        elif t == "heatmap":
            seed = int(layer["seed"])
            n = int(layer["points"])
            h = hashlib.sha256(f"heatmap-{seed}-{n}".encode()).hexdigest()[:16]
            layers_out.append(
                {
                    "type": "heatmap",
                    "digest": h,
                    "bins": list(layer["bins"]),
                    "n": n,
                    "seed": seed,
                }
            )
        elif t == "line_stub":
            layers_out.append({"type": "line_stub", "n": int(layer["n"])})
        elif t == "bar_stub":
            layers_out.append(
                {"type": "bar_stub", "categories": list(layer["categories"])}
            )
        else:
            raise ValueError(t)
    return {"layers": layers_out, "version": 1}


def main() -> None:
    p = argparse.ArgumentParser()
    p.add_argument("name", help="fixture file name under fixtures/, e.g. scenario_scatter_basic.json")
    args = p.parse_args()
    path = Path(__file__).resolve().parent / "fixtures" / args.name
    scene = scene_from_fixture(path)
    sys.stdout.write(json.dumps(scene, sort_keys=True, separators=(",", ":")))
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: tests/differential/python_ts_parity/build_scene.mjs
#!/usr/bin/env node
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const root = dirname(fileURLToPath(import.meta.url));

function stableStringify(obj) {
  if (obj === null || typeof obj !== "object") {
    return JSON.stringify(obj);
  }
  if (Array.isArray(obj)) {
    return `[${obj.map(stableStringify).join(",")}]`;
  }
  const keys = Object.keys(obj).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${stableStringify(obj[k])}`).join(",")}}`;
}

function sceneFromFixture(json) {
  if (json.version !== 1) throw new Error("version");
  const layersOut = [];
  for (const layer of json.layers) {
    const t = layer.type;
    if (t === "scatter") {
      const seed = layer.seed;
      const n = layer.points;
      const h = createHash("sha256")
        .update(`scatter-${seed}-${n}`)
        .digest("hex")
        .slice(0, 16);
      layersOut.push({ type: "scatter", digest: h, n, seed });
    } else if (t === "heatmap") {
      const seed = layer.seed;
      const n = layer.points;
      const h = createHash("sha256")
        .update(`heatmap-${seed}-${n}`)
        .digest("hex")
        .slice(0, 16);
      layersOut.push({
        type: "heatmap",
        digest: h,
        bins: layer.bins,
        n,
        seed,
      });
    } else if (t === "line_stub") {
      layersOut.push({ type: "line_stub", n: layer.n });
    } else if (t === "bar_stub") {
      layersOut.push({ type: "bar_stub", categories: layer.categories });
    } else throw new Error(t);
  }
  return { layers: layersOut, version: 1 };
}

const name = process.argv[2];
if (!name) {
  console.error("usage: build_scene.mjs <fixture.json>");
  process.exit(2);
}
const fixturePath = join(root, "fixtures", name);
const json = JSON.parse(readFileSync(fixturePath, "utf8"));
const scene = sceneFromFixture(json);
process.stdout.write(stableStringify(scene) + "\n");
